### PR TITLE
Revert civicrm_action_schedule.limit_to boolean column to be NULL

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
Revert the boolean field civicrm_action_schedule.limit_to to be NULL and not required, to address issues https://lab.civicrm.org/dev/core/-/issues/3464 and https://lab.civicrm.org/dev/core/-/issues/3465 

In my opinion, limit_to should be int with three options. 

Comments
----------------------------------------
ping @agileware-justin @eileenmcnaughton @MegaphoneJon @colemanw

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@monishdeb we need to merge to the 5.50 rc before we can merge to 5.49 stable - can you put up a PR against the rc too (this will be mergeable once that is merged, if no changes made to that one)

**Comment by monishdeb** [comment]
Created PR for 5.50 rc https://github.com/civicrm/civicrm-core/pull/23478

**Comment by eileenmcnaughton** [comment]
[Test Result](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/) (11 failures / +11)
[CivicrmUpgradeTest.CivicrmUpgradeTest.4_6_36-setupsh_sql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_4_6_36_setupsh_sql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.4_7_0-setupsh_sql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_4_7_0_setupsh_sql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.4_7_31-setupsh_sql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_4_7_31_setupsh_sql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_7_3-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_7_3_setupsh_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_13_3-multilingual_af_bg_en_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_13_3_multilingual_af_bg_en_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_13_5-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_13_5_setupsh_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_21_2-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_21_2_setupsh_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_27_4-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_27_4_setupsh_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_33_2-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_33_2_setupsh_mysql_bz2/)
[CivicrmUpgradeTest.CivicrmUpgradeTest.5_39_1-setupsh_mysql_bz2](https://test.civicrm.org/job/CiviCRM-Core-PR/48771/testReport/junit/(root)/CivicrmUpgradeTest/CivicrmUpgradeTest_5_39_1_setupsh_mysql_bz2/)

**Comment by totten** [comment]
@monishdeb I would suggest a second commit based on:

```
./tools/bin/scripts/set-version.php 5.49.2 --commit
```

**Comment by monishdeb** [comment]
> @monishdeb I would suggest a second commit based on:
> 
> ```
> ./tools/bin/scripts/set-version.php 5.49.2 --commit
> ```

done :)

**Comment by totten** [comment]
@monishdeb To try this, I created a 5.48 database with ~10 scheduled-reminders. Each reminder was in a different configuration status, eg

* An example with `limit_to="Neither" (null)`
* Two examples with `limit_to="Also Include" (0)` (*one using "Select Group"; another using "Choose Recipients")
* Three examples with `limit_to="Limit To" (1)` (*one using "Select Group"; another using "Choose Recipients"; another using "Soft Credit"*)
* A few examples which were initially saved as one thing (eg "Limit To") and then edited/downgraded to the simpler "Neither".

Then I used this DB snapshot with various upgrade paths, eg

* 5.48.2 => 5.49.1
* 5.48.2 => 5.49.1 => 5.49.2 (w/patch)
* 5.48.2 => 5.49.2 (w/patch)

The results are here: https://gist.github.com/totten/7f43420108d4e40c4d1aabbda1763f1b

A few issues:

* In `548-5492.txt `, rows `#11` and `#12` both got munged values of `limit_to`. I think this could be fixed with an extra guard: if it's a direct (*if `limit_to` is already nullable*), then `changeBooleanColumnLimitTo should not do anything.
* In `548-5491-5492.txt `, rows `#7`, `#8`, `#9`, `#10` got munged values of `limit_to=0`. Of course, this really happened in the 5.49.1 step, and I don't see how to fix those ones. However, perhaps we could give a notice to point people at records which should be re-considered?

**Comment by agileware-justin** [comment]
If limit_to=NULL then group_id should also be set to NULL - to prevent 

From https://lab.civicrm.org/dev/core/-/issues/3464#note_74627

If the limit_to value is 1 then remove the value of the Group ID field on save of this form.

**Comment by monishdeb** [comment]
@totten thanks for detailed report. I have addressed the first issue by adding an upgrade code for 5.49.0 and 2nd issue by adding pre and post upgrade message. 


@agileware-justin I will address this bug in the master branch PR https://github.com/civicrm/civicrm-core/pull/23487

File: CRM/Upgrade/Incremental/php/FiveFortyNine.php [hunk 1]
```diff
@@ -75,8 +82,8 @@ public function upgrade_5_49_beta1($rev): void {
    *
    * @param string $rev
    */
-  public function upgrade_5_49_2($rev) {
-    if (version_compare(CRM_Core_BAO_Domain::version(), '5.49', '>=') {
+  public function upgrade_5_49_0($rev) {
+    if (version_compare(CRM_Core_BAO_Domain::version(), '5.49.beta1', '>=')) {
```
Comment (totten) [inline CRM/Upgrade/Incremental/php/FiveFortyNine.php:86]: Well... that was worth a try... In my `r-run` over each of the upgrade paths, it never seems to return TRUE, so it always skipped the step.

I think I found a hidden feature which what this is trying to do... but still putting it through the `r-run`...

**Comment by totten** [comment]
Replaced by #23490, which starts from similar commits but adds a few more.

File: CRM/Upgrade/Incremental/php/FiveFortyNine.php (hunk 1 above)
Comment (monishdeb) [reply CRM/Upgrade/Incremental/php/FiveFortyNine.php:86]: Awesome, didn't knw that

**Pull request closed without merging**